onnxruntime/test/providers/provider_test_utils.h

```diff
@@ -529,7 +530,27 @@ class OpTester {
       OrtValue value;
       value.Init(p_tensor.release(), DataTypeImpl::GetType<Tensor>(),
                  DataTypeImpl::GetType<Tensor>()->GetDeleteFunc());
-      data.push_back(Data(NodeArg(name, &type_proto), std::move(value), optional<float>(), optional<float>(), sort_output));
+      auto node_arg = NodeArg(name, &type_proto);
+      if (p_dim_params && !(p_dim_params->empty())){
+        // If p_dim_params presents, configure node_arg's dim value based on p_dim_params, which supports symbolic dim and dim broadcast.
+        auto& dim_params=*p_dim_params;
+        onnx::TensorShapeProto new_shape;
+
+        // currently hard-code the reserved symbolic names.
+        // TODO: when the list grows longer, consider move it to a better place.
+        const static std::vector<std::string> reserved_symbolic{"batch", "seq"};
```

Comment:
1. you should use std::unordered_set here for fast lookup. 
2. why you need a reserved set for symbolic shape? having user to pass [a, 1, c] should still be allowed
3. you can check if string can be convert to int to determine whether to set is as symbolic shape or concrete shape

Reply:
Also, need to check if dim_params's concrete shape value matches with dim's value. 
Otherwise, it's an invalid test case

Reply:
The reason for using a reserved set is to avoid (3). Agreed that (3) would be more flexible, but I am trying to avoid going to the route of converting a string to int to just check if it is convertable. This is dangerous and can silently go wrong. On the other hand, having a fixed set gives us a simpler logic and less error-prone. Besides, as this is test code, I think it's ok put a hard-coded list of symbolic name.


---
In reply to: [417685952](https://github.com/microsoft/onnxruntime/pull/3743#discussion_r417685952) [](ancestors = 417685952)

Reply:
If the shape doesn't match, the test will still fail on data binding. But let's catch it ealier than later. 

---
In reply to: [417690715](https://github.com/microsoft/onnxruntime/pull/3743#discussion_r417690715) [](ancestors = 417690715)